api/types/events/events.proto

```diff
@@ -931,6 +931,97 @@ message SCP {
     string Action = 8 [ (gogoproto.jsontag) = "action" ];
 }
 
+// SFTPAttributes are file metadata sent over SFTP
+message SFTPAttributes {
+    // FileSize is file size
+    oneof FileSize { uint64 OptionalFileSize = 1 [ (gogoproto.jsontag) = "file_size" ]; }
+
+    // UID is the user owner of a file
+    oneof UID { uint32 OptionalUID = 2 [ (gogoproto.jsontag) = "uid" ]; }
+
+    // GID is the group owner of the file
+    oneof GID { uint32 OptionalGID = 3 [ (gogoproto.jsontag) = "gid" ]; }
+
+    // Permissions is the file permissions
+    oneof Permissions { uint32 OptionalPermissions = 4 [ (gogoproto.jsontag) = "permissions" ]; }
```

Comment:
To make fields optional. Protobuf v3 supports the `optional` keyword which essentially abstracts creating and using a `oneof` field from you, but gogo protobufs does not support `optional` so I have to do it manually.

SFTP attributes can be null essentially, so I made fields optional here so you can see what attributes were actually sent by the client.

Reply:
Aren't all fields `optional` in proto3?  cc @codingllama who understands protos better than I do

Reply:
That was in proto2 I believe, proto3 removed `optional` entirely then re-added it later :man_shrugging:

Reply:
`optional` and `oneof` are different beasts. Zac is correct, all proto3 fields are implicitly optional, meaning you can marshal/unmarshal a message without those fields. +1 for removing all the oneofs here.

The "TL;DR" of optional is that `required` was a terrible idea, so it was dropped in proto3. All fields are optional since. The re-adding of the keyword probably has to do with easing the migration from proto2 to proto3.

Reply:
@zmb3 yes, all fields are optional in proto3 and if not specified in the payload the default value is assumed.

@capnspacehook, I think that what you want here is to use proto wrappers for uint32 instead of `oneof`
https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/wrappers.proto#L96
They are essentially the underlying type but they support null value.

gogoproto can generate go code where a wrapper type is transformed into `*uint32` field

Reply:
If we want to distinguish no value assigned from the default value we can apply the  `TypeWrapper`  [pattern](https://pkg.go.dev/google.golang.org/protobuf/types/known/wrapperspb#UInt32Value) that is already used in our repo for [bool values](https://github.com/gravitational/teleport/blob/master/api/types/types.proto#L2029 )
with the `(gogoproto.nullable) = true,` property. That will allow in GO code check for nil value:

Reply:
@codingllama thanks for clearing that up.

I think the wrappers suggested by @tigrato will work best here. I agree that treating zero and nil the same way is ideal, but it's not possible here without losing information when clients send some but not all attributes. It wouldn't be possible to tell if a user sent a UID of 0 or didn't send one at all for instance.